google-cloud-storage/src/main/java/com/google/cloud/storage/WriteFlushStrategy.java

```diff
@@ -105,39 +111,51 @@ interface Flusher {
   private static final class FsyncEveryFlusher implements Flusher {
 
     private final ClientStreamingCallable<WriteObjectRequest, WriteObjectResponse> write;
+    private final RetryingDependencies deps;
+    private final ResultRetryAlgorithm<?> alg;
     private final String bucketName;
     private final LongConsumer sizeCallback;
     private final Consumer<WriteObjectResponse> completeCallback;
 
     private FsyncEveryFlusher(
         ClientStreamingCallable<WriteObjectRequest, WriteObjectResponse> write,
+        RetryingDependencies deps,
+        ResultRetryAlgorithm<?> alg,
         String bucketName,
         LongConsumer sizeCallback,
         Consumer<WriteObjectResponse> completeCallback) {
       this.write = write;
+      this.deps = deps;
+      this.alg = alg;
       this.bucketName = bucketName;
       this.sizeCallback = sizeCallback;
       this.completeCallback = completeCallback;
     }
 
     public void flush(@NonNull List<WriteObjectRequest> segments) {
-
-      Observer observer = new Observer(sizeCallback, completeCallback);
-      GrpcCallContext internalContext = contextWithBucketName(bucketName);
-      ApiStreamObserver<WriteObjectRequest> write =
-          this.write.withDefaultCallContext(internalContext).clientStreamingCall(observer);
-
-      boolean first = true;
-      for (WriteObjectRequest message : segments) {
-        if (!first) {
-          message = message.toBuilder().clearUploadId().clearWriteObjectSpec().build();
-        }
-
-        write.onNext(message);
-        first = false;
-      }
-      write.onCompleted();
-      observer.await();
+      Retrying.run(
+          deps,
+          alg,
+          () -> {
+            Observer observer = new Observer(sizeCallback, completeCallback);
```

Comment:
I don't think I've come across this before; what does Observer do for us?

Reply:
In this case, `Observer` is how our client receives events from the server. We'll get an onComplete each time the server finishes processing what we're calling a "flush". Or, an onError if some failure happens.